### ISSUE TITLE
When will 3.0 be released?

### ISSUE THREAD

**Issue by wimglenn** [marshmallow-code/marshmallow#928]
Hi, first of all thanks to all the project authors and maintainers for their work with this library! It is clean and beautiful.

I'm planning to include marshmallow in a project, but am not such a fan of the 2.x series API and I'm unsure about using a beta release on a large scale in production code. I was wondering if you have a release schedule or a tentative date for 3.0 final, or could advise about what issues (if any) are blocking a release candidate? Seems v3.0 has been in beta for [over a year now](https://pypi.org/project/marshmallow/#history).

**Comment by lafrech** [comment]
Thanks for the positive feedback.

There is no target date as I know of, but we'd certainly like to release 3.0 as soon as we can.

We'd like to close issues milestoned 3.0 first: https://github.com/marshmallow-code/marshmallow/milestone/10.

**Comment by deckar01** [comment]
I am personally not a fan of building up breaking changes in a major version milestone. I think this pattern is an artifact of commercial software development with scheduled release windows. Without a hard release schedule, there is no mechanism for punting the backlog to the next major version.

Semver + upgrade guides should provide the assurance developers need to handle major releases dropping any time an improvement is backwards incompatible.

**Comment by lafrech** [comment]
I guess the point is to avoid maintaining too many branches in parallel. Marshmallow is used in many libs (at least webargs, apispec, marshmallow-sqlalchemy, etc. to mention only its close environment). If too many branches co-exist, each lib must support many of them, with burdensome if/elif/.../else sections, or drop some versions, but then one needs to find the version that is supported by all of his libs.

This said, as pointed out in OP, marshmallow 3 has been in beta state for a long while, and it would be nice to finally allow all its improvements to be used in production.

Maybe we could inspect all those 3.0 issues and postpone some of them.

**Comment by sloria** [comment]
@deckar01 While I agree that it is nice to release as often as feasible while following semver, @lafrech is spot on re: maintaining multiple release lines. marshmallow's release cadence is a bit slower than our other libs due to the number of downstream projects that rely on it.

That said, 3.0 is already a significant improvement over 2.0, and I agree that it's high time we get a stable release out. I am fine deferring some of the issues in the 3.0 milestone. @lafrech @deckar01 I delegate to your judgment on which issues to postpone.

**Comment by hardyarora** [comment]
Any update on this?

**Comment by sloria** [comment]
See https://github.com/marshmallow-code/marshmallow/milestone/10 for progress.

**Comment by fuhrysteve** [comment]
Thanks @lafrech for pointing me over here!

**Comment by fuhrysteve** [comment]
Thanks @lafrech for pointing me over here!

Maintainer of https://github.com/fuhrysteve/marshmallow-jsonschema here checking in. I'm planning on supporting v2 and v3 out of the box for awhile to help folks (including my own team) transition.

Super excited for all the new features and enhancements, but just wanted to weigh in and voice my agreement with others that maintaining multiple diverging branches for so long is an unhealthy thing for any project. I'd much rather have a stable release with the understanding that there will be some bugs that need to get shook out than live in release-candidate land forever.

Thanks @sloria and others for all the great work coordinating the v3 release of this excellent project!

**Comment by ssfdust** [comment]
Any update?I have checked all issues milestoned 3.0 have been closed.Will it be released this week?

**Comment by sloria** [comment]
Yes, the 3.0 issues are complete. We are just allowing some time for people to use the latest RC before the final release. If you'd like to help, I encourage you to upgrade to 3.0.0rc9 and report any issues that you find.

**Comment by fuhrysteve** [comment]
@sloria we've all got your back when you're ready to pull the trigger! We're all professionals here, and know how to rollback if there are issues.

I already have it on my head that it's going to take my team 6 months after the 3.0 official release before we'll be ready to upgrade anyways (I manage a large codebase with thousands of marshmallow schemas). I'm just eager for that clock to start ticking!!

**Comment by sloria** [comment]
@fuhrysteve Start that clock: v3 is released!

**Comment by fuhrysteve** [comment]
Wow!!! :champagne: :champagne: :champagne: 

Congrads on completing such an incredible milestone. Introducing major BC breaks so late in a large and widely used library's life-cycle is no a small achievement. Myself and my team alone have spent thousands of hours of our lives working with marshmallow, and the library has had a major impact on the way I think about software architecture after so many years of using it.

We are all super grateful for the work that you and all the other contributors have put into making this happen!!

**Comment by antgel** [comment]
Sorry (not sorry) for the spam, but woo! Congrats to all, I've only been following closely for a few months, but it's an incredible effort.
